rfcs/20231025-reusable-kernels/README.md

```diff
@@ -0,0 +1,106 @@
+# Proposal to add a reusable kernel knob
+
+## Introduction
+
+Primitive creation is an expensive operation, which is why oneDNN has a
+primitive cache, a persistent cache, and a kernel cache (for the GPU). For GPU
+implementations, the kernel cache is optional and must be opted into. Although
+compiling kernels for a given problem shape can open up optimization potential,
+it comes at the cost of being incompatible with the kernel cache because the
+kernel cannot be reused for different problem shapes. Therefore, it can be
+helpful to the user to be able to specify if they prefer to use a kernel that
+has been designed to be reusable via the kernel cache (accepting any possible
+performance impacts associated with it) or the typical performance-optimized
+kernel. This proposal is to add a knob the user can control to specify whether
+they would like reusable kernels to be dispatched preferentially over the
+standard performance-optimized kernel ordering for GPU implementations.
+
+> **Note:** Reusable kernels are defined as GPU kernels with a finite set of
+configurations, which have opted-into the kernel cache internal API. Reusable
+kernels may depend on problem shape, but only in a limited way (i.e. the
+specific shape cannot be compiled directly into the kernel).
+
+## Proposal
+
+The proposal has four potential options: using DNNL_RUNTIME_DIM_VAL, a build
+flag, an environment variable, or a primitive attribute. In any of the options,
+the default behaviour will be the current one: kernels are ordered according to
+internal heuristics, roughly in order of decreasing likelihood for optimal
+performance for the given problem.
+
+If the reusable kernel option is selected, GPU kernels designed to be reusable
+will be selected first. In the case that no reusable kernels are compatible with
+the problem, kernel selection will continue with non-reusable kernels.
+
+Option 4 (primitive attribute) is the preferred option by frameworks, as they
+don't know which dimensions should be considered dynamic at runtime. It
+provides a balance between user control, flexibility, and ease of integration.
+Compared to the build flag it's more flexible, and compared to the environment
+variable it's less opaque.
+
+### Option 1: Using DNNL_RUNTIME_DIM_VAL
```

Comment:
Using `DNNL_RUNTIME_DIM_VAL` would require a primitive to contain the entire set of the kernels, which means primitive creation would be very slow as we would have to compile all of the kernels at once. This sounds like the biggest disadvantage of the approach.

Reply:
Do we have an option to do kernel compilation in a lazy way relying on the kernel cache?

Reply:
Having a lazy kernel compilation will mean that we can no longer guarantee that the primitive creation and execution are truly separate because the lazy compilation would have to be performed at the execution time. Also, that would mean that we change the state of the primitive after its creation.

Reply:
> we would have to compile all of the kernels at once
I don't understand this point. If the user supplies one or more `DNNL_RUNTIME_DIM_VAL` values, only an implementation that supports runtime dimensions will get dispatched. The intention is that any runtime dimensions will be passed at runtime instead of being compiled into the kernel, thus making a single kernel work for the entire class of problems with changing runtime sizes.
To be clear, I don't think this solution is optimal, because it limits optimization potential (e.g. we can't rely on divisibility conditions) in addition to limitations on its practical use in frameworks. I just don't see why it would require compilation of all possible kernels at once.

Reply:
@Simonsays095, the issue here is that, depending on which dimensions are set to `DNNL_RUNTIME_DIM_VAL`, there may be multiple reusable specializations that apply. Take for example the simple gen9_eltwise implementation. There is currently a compile time parameter [with_overflow](https://github.com/intel-innersource/libraries.performance.math.onednn/blob/3e71c6667fcf150c3c60e3e29c2c4b0c70f46193/src/gpu/ocl/gen9_eltwise.cpp#L59)  that depends on the total buffer size modulo a constant. As such, we would need to compile all (or some implementation specific subset of) applicable specializations to achieve the targeted performance before dispatching problems.

Reply:
@rjoursler how would that be different with the hint vs `DNNL_RUNTIME_DIM_VAL`?

Reply:
@mgouicem, there is a significant difference in the amount of information. At a high level, we have a set of specializations and a dispatching heuristic mapping a problem to a specialization. With `DNNL_RUNTIME_VAL`, we need to cover multiple problems, and, as a consequence, the dispatcher must map the set of problems to a set of specializations. For the hint, we are just reducing the set of specializations considered by the dispatching heuristic